README.md
# NAME

UV - Perl interface to libuv

# SYNOPSIS
#hactoberfest
```perl
#!/usr/bin/env perl
use strict;
use warnings;

use UV;
use UV::Loop;

# hi-resolution time
my $hi_res_time = UV::hrtime();

# A new loop
my $loop = UV::Loop->new();

# default loop
my $loop = UV::Loop->default_loop(); # convenience singleton constructor
my $loop = UV::Loop->default(); # convenience singleton constructor

# run a loop with one of three options:
# UV_RUN_DEFAULT, UV_RUN_ONCE, UV_RUN_NOWAIT
$loop->run(); # runs with UV_RUN_DEFAULT
$loop->run(UV::Loop::UV_RUN_DEFAULT); # explicitly state UV_RUN_DEFAULT
$loop->run(UV::Loop::UV_RUN_ONCE);
$loop->run(UV::Loop::UV_RUN_NOWAIT);
```

# DESCRIPTION

This module provides an interface to [libuv](http://libuv.org). We will try to
document things here as best as we can, but we also suggest you look at the
[libuv docs](http://docs.libuv.org) directly for more details on how things
work.

Event loops that work properly on all platforms. YAY!

# CONSTANTS

## VERSION CONSTANTS

### UV\_VERSION\_MAJOR

### UV\_VERSION\_MINOR

### UV\_VERSION\_PATCH

### UV\_VERSION\_IS\_RELEASE

### UV\_VERSION\_SUFFIX

### UV\_VERSION\_HEX

## ERROR CONSTANTS

### UV\_E2BIG

Argument list too long

### UV\_EACCES

Permission denied

### UV\_EADDRINUSE

Address already in use

### UV\_EADDRNOTAVAIL

Address not available

### UV\_EAFNOSUPPORT

Address family not supported

### UV\_EAGAIN

Resource temporarily unavailable

### UV\_EAI\_ADDRFAMILY

Address family not supported

### UV\_EAI\_AGAIN

Temporary failure

### UV\_EAI\_BADFLAGS

Bad ai\_flags value

### UV\_EAI\_BADHINTS

Invalid value for hints

### UV\_EAI\_CANCELED

Request canceled

### UV\_EAI\_FAIL

Permanent failure

### UV\_EAI\_FAMILY

ai\_family not supported

### UV\_EAI\_MEMORY

Out of memory

### UV\_EAI\_NODATA

No address

### UV\_EAI\_NONAME

Unknown node or service

### UV\_EAI\_OVERFLOW

Argument buffer overflow

### UV\_EAI\_PROTOCOL

Resolved protocol is unknown

### UV\_EAI\_SERVICE

Service not available for socket type

### UV\_EAI\_SOCKTYPE

Socket type not supported

### UV\_EALREADY

Connection already in progress

### UV\_EBADF

Bad file descriptor

### UV\_EBUSY

Resource busy or locked

### UV\_ECANCELED

Operation canceled

### UV\_ECHARSET

Invalid Unicode character

### UV\_ECONNABORTED

Software caused connection abort

### UV\_ECONNREFUSED

Connection refused

### UV\_ECONNRESET

Connection reset by peer

### UV\_EDESTADDRREQ

Destination address required

### UV\_EEXIST

File already exists

### UV\_EFAULT

Bad address in system call argument

### UV\_EFBIG

File too large

### UV\_EHOSTUNREACH

Host is unreachable

### UV\_EINTR

Interrupted system call

### UV\_EINVAL

Invalid argument

### UV\_EIO

i/o error

### UV\_EISCONN

Socket is already connected

### UV\_EISDIR

Illegal operation on a directory

### UV\_ELOOP

Too many symbolic links encountered

### UV\_EMFILE

Too many open files

### UV\_EMLINK

Too many links

### UV\_EMSGSIZE

Message too long

### UV\_ENAMETOOLONG

Name too long

### UV\_ENETDOWN

Network is down

### UV\_ENETUNREACH

Network is unreachable

### UV\_ENFILE

File table overflow

### UV\_ENOBUFS

No buffer space available

### UV\_ENODEV

No such device

### UV\_ENOENT

No such file or directory

### UV\_ENOMEM

Not enough memory

### UV\_ENONET

Machine is not on the network

### UV\_ENOPROTOOPT

Protocol not available

### UV\_ENOSPC

No space left on device

### UV\_ENOSYS

Function not implemented

### UV\_ENOTCONN

Socket is not connected

### UV\_ENOTDIR

Not a directory

### UV\_ENOTEMPTY

Directory not empty

### UV\_ENOTSOCK

Socket operation on non-socket

### UV\_ENOTSUP

Operation not supported on socket

### UV\_ENXIO

No such device or address

### UV\_EOF

End of file

### UV\_EPERM

Operation not permitted

### UV\_EPIPE

Broken pipe

### UV\_EPROTO

Protocol error

### UV\_EPROTONOSUPPORT

Protocol not supported

### UV\_EPROTOTYPE

Protocol wrong type for socket

### UV\_ERANGE

Result too large

### UV\_EROFS

Read-only file system

### UV\_ESHUTDOWN

Cannot send after transport endpoint shutdown

### UV\_ESPIPE

Invalid seek

### UV\_ESRCH

No such process

### UV\_ETIMEDOUT

Connection timed out

### UV\_ETXTBSY

Text file is busy

### UV\_EXDEV

Cross-device link not permitted

### UV\_UNKNOWN

Unknown error

# FUNCTIONS

The following functions are available:

## check

```perl
my $handle = UV::check(); # uses the default loop
my $handle = UV::check(loop => $some_other_loop); # non-default loop
```

Returns a new [UV::Check](https://metacpan.org/pod/UV%3A%3ACheck) Handle object.

## default\_loop

```perl
my $loop = UV::default_loop();
# You can also get it with the UV::Loop methods below:
my $loop = UV::Loop->default_loop();
my $loop = UV::Loop->default();
# Passing a true value as the first arg to the UV::Loop constructor
# will also return the default loop
my $loop = UV::Loop->new(1);
```

Returns the default loop (which is a singleton object). This module already
creates the default loop and you get access to it with this method.

## err\_name

```perl
my $error_name = UV::err_name(UV::UV_EAI_BADFLAGS);
say $error_name; # EAI_BADFLAGS
```

The [err\_name](http://docs.libuv.org/en/v1.x/errors.html#c.uv_err_name)
function returns the error name for the given error code. Leaks a few bytes of
memory when you call it with an unknown error code.

In libuv errors are negative numbered constants. As a rule of thumb, whenever
there is a status parameter, or an API functions returns an integer, a negative
number will imply an error.

When a function which takes a callback returns an error, the callback will
never be called.

## hrtime

```perl
my $uint64_t = UV::hrtime();
```

Get the current Hi-Res time (`uint64_t`).

## idle

```perl
my $handle = UV::idle(); # uses the default loop
my $handle = UV::idle(loop => $some_other_loop); # non-default loop
```

Returns a new [UV::Idle](https://metacpan.org/pod/UV%3A%3AIdle) Handle object.

## loop

```perl
my $loop = UV::loop();
# You can also get it with the UV::Loop methods below:
my $loop = UV::Loop->default_loop();
my $loop = UV::Loop->default();
```

Returns the default loop (which is a singleton object). This module already
creates the default loop and you get access to it with this method.

## poll

```perl
my $handle = UV::poll(); # uses the default loop
my $handle = UV::poll(loop => $some_other_loop); # non-default loop
```

Returns a new [UV::Poll](https://metacpan.org/pod/UV%3A%3APoll) Handle object.

## prepare

```perl
my $handle = UV::prepare(); # uses the default loop
my $handle = UV::prepare(loop => $some_other_loop); # non-default loop
```

Returns a new [UV::Prepare](https://metacpan.org/pod/UV%3A%3APrepare) Handle object.

## strerror

```perl
my $error = UV::strerror(UV::UV_EAI_BADFLAGS);
say $error; # bad ai_flags value
```

The [strerror](http://docs.libuv.org/en/v1.x/errors.html#c.uv_strerror)
function returns the error message for the given error code. Leaks a few bytes
of memory when you call it with an unknown error code.

In libuv errors are negative numbered constants. As a rule of thumb, whenever
there is a status parameter, or an API functions returns an integer, a negative
number will imply an error.

When a function which takes a callback returns an error, the callback will
never be called.

## timer

```perl
my $timer = UV::timer(); # uses the default loop
my $timer = UV::timer(loop => $some_other_loop); # non-default loop
```

Returns a new [UV::Timer](https://metacpan.org/pod/UV%3A%3ATimer) object.

## version

```perl
my $int = UV::version();
```

The [version](http://docs.libuv.org/en/v1.x/version.html#c.uv_version) function
returns `UV::UV_VERSION_HEX`, the libuv version packed into a single integer.
8 bits are used for each component, with the patch number stored in the 8 least
significant bits. E.g. for libuv 1.2.3 this would be `0x010203`.

## version\_string

```
say UV::version_string();
# 1.13.1
```

The [version\_string](http://docs.libuv.org/en/v1.x/version.html#c.uv_version_string)
function returns the libuv version number as a string. For non-release versions
the version suffix is included.

# AUTHOR

Chase Whitener <`capoeirab@cpan.org`>

# AUTHOR EMERITUS

Daisuke Murase <`typester@cpan.org`>

# COPYRIGHT AND LICENSE

Copyright 2012, Daisuke Murase.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Revision history for Perl extension UV

1.912 2021-07-25
    - Documentation adjustment preparing for 2.000 release
    - Misc small internal fixes to Build.PL and git metadata
    - Bugfix the github CI workflows
    - Add repository metadata to META files
    - Regenerated README.md
    - Back-compat to perl 5.14; declare minimum required perl version

1.911 2021-07-16
    - Fix an `#if` that should have been an `#ifdef`
    - Fix skip message in t/12-udp-open.t

1.910 2021-07-16
    - Don't attempt to support UV::TCP->open or UV::UDP->open on Windows,
      because we'd need WSA_FLAG_OVERLAPPED but Perl doesn't set it on sockets
      it creates (related to #38).

1.909 2021-05-26
    - Avoid use-after-free bug when ->close'ing newly listened client
      connections
    - Account for different errno value on MSWin32 in unit test

1.908 2021-05-18
    - Don't attempt to unlink $path at END time if the entire test was skipped
    - A better implementation of pipepair for testing that might work on
      MSWin32

1.907 2021-04-09
    - Various small fixes from @Corion to improve unit testing on MSWin32

1.906 2021-03-25
    - Don't worry about `AI_V4MAPPED` on systems which don't provide it
    - Removed now-dead `cpanfile`

1.905 TRIAL 2021-03-23
    - Various blind stabbings-in-the-dark that might be MSWin32 bugfixes
       + Export SIG* constants because libuv emulates a few
       + Handle CRLF linefeeds in Process tests
       + Maybe we need to _MAKE_SOCK() when calling uv_tcp_open() and
         uv_udp_open()
       + Skip some AF_UNIX tests that won't work there
    - Return hrtime() as an NV on platforms without 64bit int (thanks ilmari)

1.904 TRIAL 2021-03-19
    - Use cmp_ok() on unreliably-failing timer test so when it fails we at
      least get a log of the numbers
    - No need for _cleanup_loop() in unit tests
    - Remove UV::Loop->walk and related code because it isn't implemented
      and can't easily be made to work

1.903 TRIAL 2021-01-19
    - Switch build system to Module::Build::Using::PkgConfig to see if the
      late-alien handling there works better for users to install

1.902     2021-01-19
    - Adjust Makefile.PL for better pkg-config or Alien handling

1.900, 1.901
    - Totally rewritten XS logic by Paul "LeoNerd" Evans

1.000009  2019-02-07
    - Improve the way we create constants. (xenu: pr/27)
    - Update the build to use @Starter v3
    - Automate version numbers using DZil rather than incrementing by hand
    - Generate GitHub Flavored Markdown for the README now
    - Expand testing to more versions of Perl

1.000008  2018-04-21
    - Fix bug in call to Newx(). (Jim Keenan: issues/24)
    - Update UV::Loop documentation to better cover the close method.
    - No longer need Devel::GlobalDestruction.
    - First real release.

1.000007  2018-04-20 (TRIAL)
    - Fixed typo in test suite
    - Added in Test::CleanNamespaces
    - Consolidate down to one BOOT block
    - Fix a bug when removing structs from objects
    - Add some debug statements when built with PERL_UV_DEBUG environment var
      set
    - Loop destructor should just call ->close()
    - Simplify the loop singleton


1.000006  2018-04-16 (TRIAL)
    - Removed is_closing()
    - Removed is_active()
    - Removed loop_configure()
    - Removed loop_alive()
    - Removed reference to UV::Poll->new_socket() as that's all done in the
      standard ->new() constructor now.
    - Updated documentation for the Handle classes to better indicate their
      use.
    - Removed extra library detection for windows as this is now done in
      Alien::libuv
    - Bumped the requirement of Alien::libuv to 1.000
    - Removed some bad Poll tests
    - Add try/catch block for all Handle start methods
    - Removed Moo requirements

1.000005  2018-01-28 (TRIAL)
    - Update 02-loop-alive.t to more closely resemble libuv's
    - Remove lib/p5uv_handles.h (no longer needed)
    - Remove lib/p5uv_loops.h (no longer needed)
    - Ensure all Handle objects call ->_destruct() on destruction rather than
      just calling close.
    - Provide the current "closed" state of the handle to its destructor to
      assist with knowing when to free safely.
    - Rename Loop->_destroy() to Loop->_destruct() to be consistent
    - Make Loop->_destroy() use p5uv_destroy_loop() to free properly
    - A few changes at the suggestion of Devel::PPPort
    - Added a few more tests to ensure we can export things correctly
    - Added a UV::check() function. A convenience wrapper to UV::Check->new
    - Added a UV::idle() function. A convenience wrapper to UV::Idle->new
    - Added a UV::poll() function. A convenience wrapper to UV::Poll->new
    - Added a UV::prepare() function. A convenience wrapper to UV::Prepare->new
    - Reworked the tests to exercise the above convenience wrappers
    - Removed a few superfluous 'use' statements

1.000004  2018-01-25 (TRIAL)
    - Require at least ExtUtils::MakeMaker v7.12 for XSMULTI
    - Move all *.[ch] into lib with the module using XSMULTI (mohawk)
    - Redesign everything
    - Now using XS::Object::Magic to hide the C structs within our Perl objects
    - Now using Moo
    - Doing more of the work in Perl-land
    - Pretty much a complete-rewrite from v1.000003

1.000003  2017-10-26 (TRIAL)
    - Don't store Loop stashes globally (thread safety)
    - Moved constant declarations / some de-dupe efforts (Thanks Paul Evans)
    - Added Perl context to all functions that call the Perl API
    - Separated loop and handle functions from the XS interface

1.000002  2017-09-24 (TRIAL)
    - Don't store Handle stashes globally for Handle types (thread safety)
    - Simplified the typemap
    - Removed UV::default_loop from UV.xs and added to UV.pm instead
    - Renamed some macros to make more sense.
    - Removed unused macros
    - Default value for SIGPROF as it doesn't exist in Windows (useful in *nix)

1.000001  2017-08-17 (TRIAL)
    - Update some prereqs

Makefile.PL
# This Makefile.PL for UV was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.48.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use ExtUtils::MakeMaker;

use Alien::Base::Wrapper ();

my %xsbuild;

my $have_libuv_by_pkgconfig = defined eval {
    require ExtUtils::PkgConfig;
    ExtUtils::PkgConfig->atleast_version( "libuv", "1.32" );
};

my %extra_buildreq;
my %extra_prereq;

if( $have_libuv_by_pkgconfig ) {
    my %info = ExtUtils::PkgConfig->find( "libuv" );
    print STDERR "Using libuv $info{modversion} from pkg-config\n";

    my @cflags = split m/ /, $info{cflags} // "";
    my @libs   = split m/ /, $info{libs};

    %xsbuild = (
        INC  => join( " ", grep { m/^-I/ } @cflags ),
        LIBS    => [          grep {  m/-l/i } @libs ],
        LDFLAGS => join( " ", grep { !m/-l/  } @libs ),
    );
}
else {
    print STDERR "Using libuv from Alien::libuv\n";

    %xsbuild = Alien::Base::Wrapper->new('Alien::libuv')->mm_args2;
    $extra_buildreq{"Alien::Base::Wrapper"} = 0;
    $extra_prereq{"Alien::libuv"} = "1.000";
}

# Our cpanfile contains the proper configure requires already
delete $xsbuild{CONFIGURE_REQUIRES};

# We keep our XS file in the right place in lib/
$xsbuild{XSMULTI} = 1;

# Append our hax/ directory
$xsbuild{INC} .= " -Ihax";

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface to libuv",
  "AUTHOR" => "Chase Whitener <capoeirab\@cpan.org>, Daisuke Murase <typester\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Config" => 0,
    "ExtUtils::MakeMaker" => "7.12",
    %extra_buildreq,
    %extra_prereq,
  },
  "CONFIGURE_REQUIRES" => {
    "Alien::Base::Wrapper" => 0,
    "Alien::libuv" => "1.000",
    "ExtUtils::MakeMaker" => "7.12"
  },
  "DISTNAME" => "UV",
  "LICENSE" => "perl",
  "NAME" => "UV",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => "5.57",
    "Scalar::Util" => 0,
    "XSLoader" => "0.14",
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0,
    %extra_prereq,
  },
  "TEST_REQUIRES" => {
    "Errno" => 0,
    "ExtUtils::MakeMaker" => "7.12",
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IO::Socket::INET" => 0,
    "POSIX" => 0,
    "Socket" => 0,
    "Test::More" => "0.88",
    %extra_prereq,
  },
  "VERSION" => "1.902",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    %xsbuild,
);

my %FallbackPrereqs = (
  "Alien::Base::Wrapper" => 0,
  "Carp" => 0,
  "Config" => 0,
  "Errno" => 0,
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => "7.12",
  "File::Spec" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "IO::Socket::INET" => 0,
  "POSIX" => 0,
  "Scalar::Util" => 0,
  "Socket" => 0,
  "Test::More" => "0.88",
  "XSLoader" => "0.14",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0,
  %extra_prereq,
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
